### PR TITLE
Spherical chunk geometrical model

### PR DESCRIPTION
**Pull request by bobmyhill**
Opened for comments. 
Plan to add oblateness (to an ellipsoidal chunk model). 

Still need to add output for tests.

### CODE REVIEW

**Comment by tjhei** [comment]
yeah, tests are not doing anything right now. :-)

**Comment by tjhei** [comment]
I don't like the name SphericalChunk because it isn't spherical.

File: include/aspect/geometry_model/spherical_chunk.h [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+namespace aspect
+{
+  namespace GeometryModel
+  {
+    using namespace dealii;
+
+    /**
+     * A class that describes a box geometry of certain width, height, and
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_chunk.h]: comment is out of date

File: include/aspect/geometry_model/spherical_chunk.h [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        /**
+               * Maximum depth, longitude-depth or
+         * longitude-latitude-depth point
+               */
+        Point<dim> point2;
+
+        /**
+         * Flag whether the box is periodic in the x-, y-, and z-direction.
+         */
+        bool periodic;
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_chunk.h]: what about periodicity in some but not all directions? Does periodicity even work?

File: include/aspect/geometry_model/spherical_chunk.h [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+         * The number of cells in each coordinate direction
+         */
+        unsigned int repetitions[dim];
+
+        /**
+         * Pointer to an object that describes the geometry.
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_chunk.h]: not a pointer, so update the comment

File: include/aspect/geometry_model/spherical_chunk.h [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+         */
+        SphericalChunkGeometry<dim> manifold;
+
+        static void set_manifold_ids (Triangulation<dim> &triangulation)
```
Comment (tjhei) [inline include/aspect/geometry_model/spherical_chunk.h]: Can you move those functions into the .cc?

File: source/geometry_model/spherical_chunk.cc [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+      {
+        case 2:
+        {
+          output_vertex[0] = std::atan2(v[0], v[1]);
+          output_vertex[1] = std::sqrt(v[0]*v[0] + v[1]*v[1]);
+          break;
+        }
+        case 3:
+        {
+          double radius=std::sqrt(v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: make radius const

Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: and use v.norm() instead.

File: source/geometry_model/spherical_chunk.cc [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+  SphericalChunkGeometry<dim>::
+  pull_back(const Point<dim> &v) const
+  {
+    Point<dim> output_vertex;
+    switch (dim)
+      {
+        case 2:
+        {
+          output_vertex[0] = std::atan2(v[0], v[1]);
+          output_vertex[1] = std::sqrt(v[0]*v[0] + v[1]*v[1]);
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: use v.norm()

File: source/geometry_model/spherical_chunk.cc [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+                                                 point1,
+                                                 point2,
+                                                 true);
+
+      //Tell p4est about the periodicity of the mesh.
+      std::vector<GridTools::PeriodicFacePair<typename parallel::distributed::Triangulation<dim>::cell_iterator> >
+      periodicity_vector;
+      if (periodic)
+        GridTools::collect_periodic_faces
+        ( coarse_grid, /*b_id1*/ 0, /*b_id2*/ 1,
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: did you check that this works after you "bend" the box?

File: source/geometry_model/spherical_chunk.cc [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+           cell != coarse_grid.end(); ++cell)
+        for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+          if (cell->face(f)->at_boundary())
+            cell->face(f)->set_all_boundary_indicators(f);
+
+      // Attach the real manifold to slot 15. we won't use it
+      // during regular operation, but we set manifold_ids for all
+      // cells, faces and edges immediately before refinement and
+      // clear it again afterwards
+      coarse_grid.set_manifold (15, manifold);
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: why do you set/unset the manifold ids?

File: source/geometry_model/spherical_chunk.cc [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+
+
+    template <int dim>
+    std::set< std::pair< std::pair<types::boundary_id, types::boundary_id>, unsigned int> >
+    SphericalChunk<dim>::
+    get_periodic_boundary_pairs () const
+    {
+      std::set< std::pair< std::pair<types::boundary_id, types::boundary_id>, unsigned int> > periodic_boundaries;
+      if (periodic)
+        periodic_boundaries.insert( std::make_pair( std::pair<types::boundary_id, types::boundary_id>(0, 1), 0) );
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: this looks wrong to me.

File: source/geometry_model/spherical_chunk.cc [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+    double
+    SphericalChunk<dim>::
+    length_scale () const
+    {
+      // as described in the first ASPECT paper, a length scale of
+      // 10km = 1e4m works well for the pressure scaling for earth
+      // sized spherical shells. use a length scale that
+      // yields this value for the R0,R1 corresponding to earth
+      // but otherwise scales like (R1-R0)
+      return 1e4 * maximal_depth() / (6336000.-3481000.);
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: is this still appropriate?

File: source/geometry_model/spherical_chunk.cc [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+    double
+    SphericalChunk<dim>::depth(const Point<dim> &position) const
+    {
+      return std::min (std::max (point2[dim-1]-position.norm(), 0.), maximal_depth());
+    }
+
+
+    template <int dim>
+    Point<dim>
+    SphericalChunk<dim>::representative_point(const double depth) const // NOT DONE YET XXXX
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: I would use the half-way point (middle) in long/lat direction.

File: source/geometry_model/spherical_chunk.cc [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+        {
+
+          const double degtorad = dealii::numbers::PI/180;
+
+          Assert (dim >= 1, ExcInternalError());
+          Assert (dim <= 3, ExcInternalError());
+
+          if (dim == 1)
+            {
+              point1[0] = prm.get_double ("Chunk bottom radius");
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: you don't need dim=1

File: source/geometry_model/spherical_chunk.cc [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+              point2[0] = prm.get_double ("Chunk top radius");
+              repetitions[0] = prm.get_integer ("Radius repetitions");
+            }
+
+          if (dim == 2)
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: run through this in 2d and 3d and only add point*[2] if dim=3

File: source/geometry_model/spherical_chunk.cc [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+                             "Radius at the top surface of the chunk. Units: m.");
+
+          prm.declare_entry ("Chunk minimum longitude", "0",
+                             Patterns::Double (-180, 180),
+                             "Minimum longitude of the chunk. Units: m.");
+          prm.declare_entry ("Chunk maximum longitude", "1",
+                             Patterns::Double (-180, 180),
+                             "Maximum longitude of the chunk. Units: m.");
+
+          prm.declare_entry ("Chunk minimum latitude", "0",
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: only do these if dim==3

File: source/geometry_model/spherical_chunk.cc [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+namespace aspect
+{
+  namespace GeometryModel
+  {
+    ASPECT_REGISTER_GEOMETRY_MODEL(SphericalChunk,
+                                   "spherical chunk",
+                                   "A box-like geometry along lines of longitude, latitude and radius. "
+                                   "The extent of the box in each coordinate direction "
+                                   "is set in the parameter file. The box geometry labels its "
+                                   "2*dim sides as follows: in 2d, boundary indicators 0 through 3 "
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: remove the numbers for the indicators. We have names for them now.

File: source/geometry_model/spherical_chunk.cc [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+                                   "denote the minimum and maximum longitude values and the "
+                                   "minimum and maximum radii; in 3d, boundary indicators 0 through 5 "
+                                   "indicate the minimum and maximum longitude, minimum and maximum "
+                                   "latitude and the minimum and maximum radius boundaries "
+                                   "(see also the documentation of the deal.II class ``GeometryInfo''). "
+                                   "You can also use symbolic names ``left'', ``right'', "
+                                   "etc., to refer to these boundaries in input files.")
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: can you list all names please?

File: source/geometry_model/spherical_chunk.cc [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+
+
+      // Transform box into spherical chunk
+      GridTools::transform (std_cxx11::bind(&SphericalChunkGeometry<dim>::push_forward,
+                                            std_cxx11::cref(manifold),
+                                            std_cxx11::_1),
+                            coarse_grid);
+
+      // Deal with a curved mesh
+      // Loop over cells,faces and set boundary indicators
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: aren't those already correct from subdivided_hyper_rectangle?

File: include/aspect/geometry_model/spherical_chunk.h (hunk 2 above)
Comment (tjhei) [inline include/aspect/geometry_model/spherical_chunk.h]: update the comment...

File: source/geometry_model/spherical_chunk.cc (hunk 8 above)
Comment (bangerth) [inline source/geometry_model/spherical_chunk.cc]: This is copied from the spherical shell. We need the manifold for mesh refinement, but manifolds do not provide normal vectors for the Q4 mapping, so we remove the manifold and fall back to the StraightBoundary object during regular operation.

File: source/geometry_model/spherical_chunk.cc [hunk 18]
```diff
@@ -0,0 +1,416 @@
...
+              point2[0] = prm.get_double ("Chunk maximum longitude") * degtorad;
+              repetitions[0] = prm.get_integer ("Longitude repetitions");
+              point1[1] = prm.get_double ("Chunk minimum latitude") * degtorad;
+              point2[1] = prm.get_double ("Chunk maximum latitude") * degtorad;
+              repetitions[1] = prm.get_integer ("Latitude repetitions");
+              point1[2] = prm.get_double ("Chunk bottom radius");
+              point2[2] = prm.get_double ("Chunk top radius");
+              repetitions[2] = prm.get_integer ("Radius repetitions");
+            }
+
```
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: can you add some checks that min(long or lat) < max(long or lat) please?

File: source/geometry_model/spherical_chunk.cc (hunk 9 above)
Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: oh, okay. Please add the comment that it is only periodic in lat if it works.

File: source/geometry_model/spherical_chunk.cc (hunk 10 above)
Comment (bobmyhill) [inline source/geometry_model/spherical_chunk.cc]: Should be.

File: source/geometry_model/spherical_chunk.cc (hunk 13 above)
Comment (bobmyhill) [inline source/geometry_model/spherical_chunk.cc]: But the order of inputs is different ... in 2D 
point*[1] =prm.get_double ("Chunk * radius");
in 3D:
point*[1] =prm.get_double ("Chunk * latitude");

Comment (tjhei) [inline source/geometry_model/spherical_chunk.cc]: oh, you are right. I guess one could do point[dim-1] = ... raduis... though.

**Comment by bobmyhill** [comment]
Ok, I've reworked the model to convert between [r, phi, theta] and [x, y, z] in a reasonable way. The prm files still take latitude and longitude (in degrees) as inputs.

**Comment by bobmyhill** [comment]
Do not merge this! Merging Pull request #478 will do everything correctly.

**Comment by gassmoeller** [comment]
Closed by #478

**Pull request closed without merging**